service/controller/v22/resource/s3bucket/delete.go

```diff
@@ -23,62 +35,92 @@ func (r *Resource) ApplyDeleteChange(ctx context.Context, obj, deleteChange inte
 		return microerror.Mask(err)
 	}
 
-	for _, bucketInput := range bucketsInput {
-		if bucketInput.Name == "" {
-			continue
-		}
+	g, ctx := errgroup.WithContext(ctx)
 
-		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("deleting S3 bucket %q", bucketInput.Name))
+	for _, b := range bucketsInput {
+		bucketName := b.Name
 
-		var repeat bool
-		for {
-			i := &s3.ListObjectsV2Input{
-				Bucket: aws.String(bucketInput.Name),
-			}
-			o, err := sc.AWSClient.S3.ListObjectsV2(i)
-			if err != nil {
-				return microerror.Mask(err)
-			}
-			if o.IsTruncated != nil && *o.IsTruncated {
-				repeat = true
-			}
-			if len(o.Contents) == 0 {
-				break
+		g.Go(func() error {
+			if bucketName == "" {
+				return nil
 			}
 
-			r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("deleting %d objects", len(o.Contents)))
+			r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("deleting S3 bucket %q", bucketName))
 
-			for _, o := range o.Contents {
-				i := &s3.DeleteObjectInput{
-					Bucket: aws.String(bucketInput.Name),
-					Key:    o.Key,
+			var count int
+			var repeat bool
+			for {
+				i := &s3.ListObjectsV2Input{
+					Bucket: aws.String(bucketName),
 				}
-				_, err := sc.AWSClient.S3.DeleteObject(i)
+				o, err := sc.AWSClient.S3.ListObjectsV2(i)
 				if err != nil {
 					return microerror.Mask(err)
 				}
-			}
+				if o.IsTruncated != nil && *o.IsTruncated {
+					repeat = true
+				}
+				if len(o.Contents) == 0 {
+					break
+				}
 
-			if !repeat {
-				break
-			}
-		}
+				r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("deleting %d objects", len(o.Contents)))
+
+				for _, o := range o.Contents {
+					i := &s3.DeleteObjectInput{
+						Bucket: aws.String(bucketName),
+						Key:    o.Key,
+					}
+					_, err := sc.AWSClient.S3.DeleteObject(i)
+					if err != nil {
+						return microerror.Mask(err)
+					}
+				}
+
+				count++
+				if count >= loopLimit {
+					r.logger.LogCtx(ctx, "level", "debug", "message", "loop limit reached for S3 bucket deletion")
+
+					r.logger.LogCtx(ctx, "level", "debug", "message", "canceling S3 bucket deletion")
+					repeat = false
+				}
 
-		{
-			i := &s3.DeleteBucketInput{
-				Bucket: aws.String(bucketInput.Name),
+				if !repeat {
+					break
+				}
 			}
 
-			_, err = sc.AWSClient.S3.DeleteBucket(i)
-			if IsBucketNotFound(err) {
-				// Fall through.
-				return nil
-			} else if err != nil {
-				return microerror.Mask(err)
+			{
+				i := &s3.DeleteBucketInput{
```

Comment:
When bucket deletion is not allowed as long as it's not empty, I'd like to have a check for that in here (we already check for empty listing above) and avoid unnecessary error below. It somehow feels wrong to me to try brute force something when we already know won't succeed.

Reply:
I don't know how to do this. The delegation along the code trace is straight forward. Everything else I came up with was worse. Maybe create a PR and show me how to do it.

Reply:
What about this: https://github.com/giantswarm/aws-operator/pull/1371 something I didn't notice there?

Reply:
It is in.